pkg/image/controller/controller.go

```diff
@@ -1,32 +1,63 @@
 package controller
 
 import (
-	"fmt"
-	"time"
-
 	"github.com/golang/glog"
 
 	kapi "k8s.io/kubernetes/pkg/api"
-	"k8s.io/kubernetes/pkg/api/errors"
-	"k8s.io/kubernetes/pkg/api/unversioned"
-	kerrors "k8s.io/kubernetes/pkg/util/errors"
-	"k8s.io/kubernetes/pkg/util/sets"
 
 	"github.com/openshift/origin/pkg/client"
-	"github.com/openshift/origin/pkg/dockerregistry"
 	"github.com/openshift/origin/pkg/image/api"
 )
 
 type ImportController struct {
-	streams  client.ImageStreamsNamespacer
-	mappings client.ImageStreamMappingsNamespacer
-	// injected for testing
-	client dockerregistry.Client
+	streams client.ImageStreamsNamespacer
+}
+
+// tagImportable is true if the given TagReference is importable by this controller
+func tagImportable(tagRef api.TagReference) bool {
+	if tagRef.From == nil {
+		return false
+	}
+	if tagRef.From.Kind != "DockerImage" || tagRef.Reference {
+		return false
+	}
+	return true
+}
+
+// tagNeedsImport is true if the observed tag generation for this tag is older than the
+// specified tag generation (if no tag generation is specified, the controller does not
+// need to import this tag).
+func tagNeedsImport(stream *api.ImageStream, tag string, tagRef api.TagReference, importWhenGenerationNil bool) bool {
+	if !tagImportable(tagRef) {
+		return false
+	}
+	if tagRef.Generation == nil {
+		return importWhenGenerationNil
+	}
+	return *tagRef.Generation > api.LatestObservedTagGeneration(stream, tag)
 }
 
 // needsImport returns true if the provided image stream should have its tags imported.
-func needsImport(stream *api.ImageStream) bool {
-	return stream.Annotations == nil || len(stream.Annotations[api.DockerImageRepositoryCheckAnnotation]) == 0
+func needsImport(stream *api.ImageStream) (ok bool, partial bool) {
```

Comment:
If you have a method that defines boolean ("IsFoo" or "needsImport") then "ok" is our convention for our answer variable.  I'll update the godoc to describe partial.